gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -0,0 +1,820 @@
+   * proper, we open the fd using udisks so we can write to it as an
+   * unpriveleged user, but also need to be able to run a script on the device,
+   * elevated with pkexec.
+   *
+   * Why not accept a UDisksBlock and perform this step internally? It's
+   * convenient for testing to be able to operate on a regular file, too.
+   */
+  props[PROP_DRIVE_PATH] = g_param_spec_string (
+      "drive-path",
+      "Drive path",
+      "Path to target drive.",
+      NULL,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  props[PROP_DRIVE_FD] = g_param_spec_int (
+      "drive-fd",
+      "Drive FD",
+      "Writable file descriptor for drive-path, which is guaranteed to be "
+      "close()d by this class.",
+      -1, G_MAXINT, -1,  /* -1 for "invalid" */
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  props[PROP_CONVERT_TO_MBR] = g_param_spec_boolean (
+      "convert-to-mbr",
+      "Convert to MBR?",
+      "Whether to convert the partition table from GPT to MBR after writing",
+      FALSE,
+      G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
+
+  props[PROP_STEP] = g_param_spec_uint (
+      "step",
+      "Step",
+      "Current step, indexed from 1",
+      1, 2, 1,
+      G_PARAM_READABLE | G_PARAM_STATIC_STRINGS);
+
+  props[PROP_PROGRESS] = g_param_spec_double (
+      "progress",
+      "Progress",
+      "Progress in the current step, between 0 and 1 inclusive, "
+      "or -1 if exact progress can't be determined.",
+      -1, 1, 0,
+      G_PARAM_READABLE | G_PARAM_STATIC_STRINGS);
+
+  g_object_class_install_properties (object_class, N_PROPERTIES, props);
+}
+
+static void
+gis_scribe_init (GisScribe *self)
+{
+  g_mutex_init (&self->mutex);
+  self->drive_fd = -1;
+  self->step = 1;
+}
+
+GisScribe *
+gis_scribe_new (GFile       *image,
+                guint64      image_size,
+                GFile       *signature,
+                const gchar *drive_path,
+                gint         drive_fd,
+                gboolean     convert_to_mbr)
+{
+  g_return_val_if_fail (G_IS_FILE (image), NULL);
+  g_return_val_if_fail (image_size > 0, NULL);
+  g_return_val_if_fail (G_IS_FILE (signature), NULL);
+  g_return_val_if_fail (drive_path != NULL, NULL);
+  g_return_val_if_fail (drive_fd >= 0, NULL);
+
+  return g_object_new (
+      GIS_TYPE_SCRIBE,
+      "image", image,
+      "image-size", image_size,
+      "signature", signature,
+      "drive-path", drive_path,
+      "drive-fd", drive_fd,
+      "convert-to-mbr", convert_to_mbr,
+      NULL);
+}
+
+static gboolean
+gis_scribe_update_progress (gpointer data)
+{
+  GisScribe *self = GIS_SCRIBE (data);
+  guint64 bytes_written;
+
+  g_mutex_lock (&self->mutex);
+  bytes_written = self->bytes_written;
+  g_mutex_unlock (&self->mutex);
+
+  self->progress = ((gdouble) bytes_written) / ((gdouble) self->image_size);
+  g_object_notify_by_pspec (G_OBJECT (self), props[PROP_PROGRESS]);
```

Comment:
Since our smallest image is 7 GiB uncompressed and we write 1 MiB to disk at a time, if the bytes written (and hence progress) does not change at least once per second (the frequency of this callback) then writing the image to disk would take more than 7168 seconds, ie 2 hours. It doesn't, even on a slow spinning disk, so in practice (and notwithstanding double precision floating point accuracy) the value of `self->progress` has almost certainly changed.

Reply:
Yes, but in the absence of actual bugs in your code, I have to comment on something to make the code review worthwhile, right?